### PR TITLE
Add example for product listing

### PR DESCRIPTION
**Pull request by kul3r4**
Solves: #256 
The first [draft](https://amp-ecommerce-analyics.firebaseapp.com/Product_Listing_Sample.html) of this example didn't contain any search form logic (calling the backend for search and retrieve the values), do you think it would be nice to add it?
@sebastianbenz PTAL

### CODE REVIEW

File: src/50_Sample_AMPs/Product_Listing_Sample.html [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    </head>
+    <body>
+    <!-- #### Seller Logo -->
+    <!-- Use `amp-img` to add the logo of the seller. --> 
+    <div class="seller-logo">
+            <amp-img src="/img/google_store.png" width="163" height="35"></amp-img>
+    </div>
+    <!-- #### Search Form -->
+    <!-- Use [amp-form](https://github.com/ampproject/amphtml/blob/master/extensions/amp-form/amp-form.md0) to realize a search form.
```
Comment (sebastianbenz) [inline src/50_Sample_AMPs/Product_Listing_Sample.html]: s/realize/add/

**Comment by sebastianbenz** [comment]
* rename file s/Product_Listing_Sample.html/Product_Listing/

**Comment by sebastianbenz** [comment]
Looks good otherwise!

It would be fantastic to have a working (simple) search!

File: backend/amp-form.go [hunk 2]
```diff
@@ -60,6 +74,33 @@ func submitForm(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func submitProductListingSearch(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "post only", http.StatusMethodNotAllowed)
+		return
+	}
+
+		products:= make([]Product, 0)
```
Comment (sebastianbenz) [inline backend/amp-form.go]: make this a constant

File: src/50_Samples_%26_Templates/Product_Listing.html [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                <input type="text" class="search-input" name="Product" required>
+            </label>
+            <input type="submit" class="button button-primary search-input-submit " value="Search">
+        </fieldset>
+    </form>
+    <!-- #### Products -->
+    <!-- Use `amp-list` to show a list of products. Learn more about `amp-list` [here](/components/amp-list/) -->
+    <h3 class="title">Products</h3>
+    <ul>
+        <amp-list width=300
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Listing.html]: do we still need amp-list?

**Comment by sebastianbenz** [comment]
* can we move the server logic into its own file product-listing.go?
* if we follow Rudy's suggestion and don't use amp-list we can move everything into the backend and also don't need a separate page for the results.

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

File: backend/product-listing.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+func registerProductListingHandler(sampleType string, sampleName string) {
+	url := path.Join("/", sampleType, sampleName) + "/"
+	filePath := path.Join(DIST_FOLDER, sampleType, sampleName, "index.html")
+	http.HandleFunc(url, func(w http.ResponseWriter, r *http.Request) {
+		renderProductListingSample(w, r, filePath, products)
+	})
+}
+
+func createProduct(name string, image string, url string, stars string) Product {
```
Comment (sebastianbenz) [inline backend/product-listing.go]: change the URL for all products to go to the product page preview.

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

File: backend/product-listing.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	"net/http"
+	"path"
+	"strings"
+)
+
+const (
+	SAMPLE_TEMPLATE_FOLDER = "samples_templates"
+)
+
+type ProductPage struct {
```
Comment (sebastianbenz) [inline backend/product-listing.go]: Isn't this ProductListingPage?

File: backend/product-listing.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	Stars string
+}
+
+var products []Product
+
+func InitProductListing() {
+	products = make([]Product, 0)
+	products = append(products,
+		Product{"Nexus 5x",
+			"https://lh3.googleusercontent.com/O00xttgXZHWYdpiWi7K0y7J2kvBMRXwlumBQNLV1853AMtf7MBBMSJ_ug9MDyUxMNjyo",
```
Comment (sebastianbenz) [inline backend/product-listing.go]: Can you make each product point to the preview version of our product sample?

File: backend/product-listing.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			"https://lh3.googleusercontent.com/ysmhjWszkLsyiWRJ97JZyL0oXL0HUzKXfhOgYC8wvLHhQUwr_dRADE1tBvUNzXPGrA",
+			"https://store.google.com/product/chromecast_audio",
+			"★★★★★"},
+	)
+	registerProductListingHandler("product_listing")
+	registerProductListingHandler("product_listing/preview")
+}
+
+func registerProductListingHandler(sampleName string) {
+	route := path.Join("/", SAMPLE_TEMPLATE_FOLDER, sampleName) + "/"
```
Comment (sebastianbenz) [inline backend/product-listing.go]: Can you put the root slash in SAMPLE_TEMPLATE_FOLDER?

File: backend/product-listing.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+		handleSearchRequest(w, r, sampleName)
+	})
+}
+
+func renderProductListing(w http.ResponseWriter, r *http.Request, sampleName string) {
+	productsToShow := searchProducts(r.URL.Query().Get("search"))
+	filePath := path.Join(DIST_FOLDER, SAMPLE_TEMPLATE_FOLDER, sampleName, "index.html")
+	t, _ := template.ParseFiles(filePath)
+	w.Header().Set("Cache-Control", fmt.Sprintf("max-age=%d, public, must-revalidate", MAX_AGE_IN_SECONDS))
+	t.Execute(w, ProductPage{Products: productsToShow, SearchAction: "/samples_templates/" + sampleName + "/search"})
```
Comment (sebastianbenz) [inline backend/product-listing.go]: SearchAction: join(SAMPLE_TEMPLATE_FOLDER, sampleName, "search)

File: backend/product-listing.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	http.HandleFunc(route, func(w http.ResponseWriter, r *http.Request) {
+		renderProductListing(w, r, sampleName)
+	})
+	http.HandleFunc(route+"search", func(w http.ResponseWriter, r *http.Request) {
+		handleSearchRequest(w, r, sampleName)
+	})
+}
+
+func renderProductListing(w http.ResponseWriter, r *http.Request, sampleName string) {
+	productsToShow := searchProducts(r.URL.Query().Get("search"))
```
Comment (sebastianbenz) [inline backend/product-listing.go]: extract "search" get parameter into a constant

File: backend/product-listing.go (hunk 8 above)
Comment (sebastianbenz) [inline backend/product-listing.go]: extract "search" form path into a constant

File: src/50_Samples_%26_Templates/Product_Listing.html [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+<!doctype html>
+<html ⚡>
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+
+    <!--
+    Additionally used AMP components must be imported in the header.
+    -->
+
```
Comment (sebastianbenz) [inline src/50_Samples_%26_Templates/Product_Listing.html]: nit: remove empty line

**Comment by sebastianbenz** [comment]
View small nits - otherwise LGTM

File: backend/product-listing.go (hunk 6 above)
Comment (kul3r4) [inline backend/product-listing.go]: I don't understand this. The product sample has only one product which is a Nexus 5. What's the point of pointing all of those products to the same one?

File: backend/product-listing.go (hunk 7 above)
Comment (kul3r4) [inline backend/product-listing.go]: Line 87 (below the code) I am using that constant and having the `/` will make things difficult.

`filePath := path.Join(DIST_FOLDER, SAMPLE_TEMPLATE_FOLDER, sampleName, "index.html")`

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

File: backend/product-listing.go (hunk 7 above)
Comment (sebastianbenz) [inline backend/product-listing.go]: I think join will do the right thing.

File: backend/product-listing.go (hunk 6 above)
Comment (sebastianbenz) [inline backend/product-listing.go]: because it's an example flow

File: backend/product-listing.go (hunk 7 above)
Comment (kul3r4) [inline backend/product-listing.go]: you are right, I removed it!

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

**Comment by sebastianbenz** [comment]
LGTM

**Pull request merged**